### PR TITLE
Random Fish Spawning

### PR DESCRIPTION
**Pull request by Wealthyturtle**
### Description

Allows scripts to be able to spawn random fish from the expression ``spawn fish``

Tested with:
``!spawn cod``
``!spawn puffer fish``
``!spawn salmon``
``!spawn tropical fish`` (random functionality still works for this)
``!spawn fish`` (will now randomly spawn cod, puffer fish, salmon, or a random tropical fish!)

---
**Target Minecraft Versions:**     1.13 and above (where fishes became entities)
**Requirements:**     None
**Related Issues:**     #2046

### CODE REVIEW

File: src/main/java/ch/njol/skript/entity/FishData.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+	@Override
+	public boolean isSupertypeOf(EntityData<?> e) {
+		return e instanceof Fish ? getPattern((Fish) e) == pattern : false;
+	}
+	
+	private static int getInitPattern(@Nullable Fish f) {
+		if(f == null)
+			return 0;
+		else if(f instanceof Cod)
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/entity/FishData.java]: Add a space after the if, in all of the if statements

File: src/main/java/ch/njol/skript/entity/FishData.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	protected boolean match(Fish entity) {
+		return getPattern(entity) == pattern;
+	}
+
+	@Override
+	public Class<? extends Fish> getType() {
+		if(pattern == 1)
+			return Cod.class;
+		else if(pattern == 2)
+			return PufferFish.class;
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/entity/FishData.java]: You could use a switch here.

File: src/main/java/ch/njol/skript/entity/FishData.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		else if(pattern == 3)
+			return Salmon.class;
+		else if(pattern == 4)
+			return TropicalFish.class;
+		return Fish.class;
+	}
+
+	@Override
+	public EntityData getSuperType() {
+		return new FishData();
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/entity/FishData.java]: What does `type of` expression return when you do this to a fish? Does comparison overall work properly? Like, `if event-entity is a fish`.

**Comment by Wealthyturtle** [comment]
Using
```
on damage:
	if victim is fish:
		broadcast "FISH"
	if victim is tropical fish:
		broadcast "TROPICAL"
	if victim is cod:
		broadcast "COD"
	if victim is puffer fish:
		broadcast "PUFFER FISH"
	if victim is salmon:
		broadcast "SALMON"
```
And spawning in some fishes, the general-purpose ``if victim is fish`` will always trigger, but with the exception of ``if victim is puffer fish``, the rest of the other fish conditionals do not trigger due to comparison issues

File: src/main/java/ch/njol/skript/entity/FishData.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	@Override
+	protected boolean match(Fish entity) {
+		return wildcard ? true : getPattern(entity) == pattern;
+	}
+
+	@Override
+	public Class<? extends Fish> getType() {
+		if(!init && wildcard)
```
Comment (Matocolotoe) [inline src/main/java/ch/njol/skript/entity/FishData.java]:
```suggestion
		if (!init && wildcard)
```

**Comment by bensku** [comment]
@FranKusmiruk please merge when approved.

**Pull request merged**